Add a `max_dust_htlc_exposure_msat`

ariard opened the pull request.

niftynei commented on 02-peer-protocol.md:
```diff
@@ -1118,6 +1173,14 @@ The node _responsible_ for paying the Bitcoin fee:
+  - if the `dust_balance_on_counterparty_tx` at the new `dust_buffer_feerate` is superior to `max_dust_htlc_exposure_msat`:
+    - MAY NOT send `update_fee`
+    - MAY fail the channel
+  - if the `dust_balance_on_holder_tx` at the new `dust_buffer_feerate` is superior to the `max_dust_htlc_exposure_msat`:
```
This should be 
```suggestion
  - if the `dust_balance_on_holder_tx` at the new `feerate_per_kw` is superior to the `max_dust_htlc_exposure_msat`:
```

The real question is if the new feerate pushes the balance over the exposure limit, not if the new feerate + buffer does. Pushing a feerate update where the new buffer ceiling is above the current should just fail to add any new dusty htlcs, but will leave the existing ones without a problem. Otherwise what was the point of having the buffer in the first place?

ariard replied:
Yes your right that's the spec doesn't match our LDK code on that: https://github.com/rust-bitcoin/rust-lightning/blob/7aa2caccd884cd7d40ee146323c2a65b7ea39407/lightning/src/ln/channel.rs#L3130

> Pushing a feerate update where the new buffer ceiling is above the current should just fail to add any new dusty htlcs, but will leave the existing ones without a problem. 

Yes that's right. If the inbound/outbound feerate is above the the dust buffer feerate ceiling, the dust exposure on counterparty and holder commitment transactions should be evaluated. If one of them is above the dust HTLC exposure the channel _could_  be fail (I think a MAY is better).

We had multiple back-and-forth on the LDK-side on the design of `update_fee` reception mitigations.

Updating the PR.

ariard replied:
Updated at 6a1cb21. If you think there are more corrections needed, let me know.

niftynei commented on 02-peer-protocol.md:
imo this whole section should be moved into the `update_add_htlc` requirements, as they're actions that are triggered by the receipt of this message.

ariard replied:
Well, in fact it's not exactly clear that it should be in the `update_add_htlc` requirements as it's a relay requirement, to me in the same category of `cltv_expiry_delta` selection. At least was my thinking when I wrote those lines. 

Though it can be a source of confusion and maybe we should even add a new BOLT4 message. Due to lack of that, I remember it was hard to assess that `max_dust_htlc_exposure_msat` was well in-place when I verified fixes against Eclair in a black-box way.

niftynei replied:
yes but you don't enact these spec requirements until a `update_add_htlc` message is received.

niftynei commented on 02-peer-protocol.md:
```diff
@@ -1127,6 +1190,17 @@ A receiving node:
     - SHOULD fail the channel,
       - but MAY delay this check until the `update_fee` is committed.
+  - if the `dust_balance_on_counterparty_tx` at the new `dust_buffer_feerate` is superior to `max_dust_htlc_exposure_msat`:
+    - MAY fail the channel
```
```suggestion
    - SHOULD fail the channel
```

ariard replied:
See the rational just under. That's a MAY as automatic-closure might bring newer issues. A third-party to a channel could relay and withhold HTLCs across the link to provoke a close.

See also Eclair new config settings `close-on-update-fee-dust-exposure-overflow` : https://github.com/ACINQ/eclair/pull/1985/files, which is false by default. Though in LDK we do auto-close by default.

As it's arbitrating between two safety risks, I think it's wiser to defer the choice to the implementator.

niftynei replied:
Well 'SHOULD' is still a deferral to the implementor, just a more strongly worded form!

niftynei commented on 02-peer-protocol.md:
```diff
@@ -1127,6 +1190,17 @@ A receiving node:
+  - if the `dust_balance_on_counterparty_tx` at the new `dust_buffer_feerate` is superior to `max_dust_htlc_exposure_msat`:
+    - MAY fail the channel
+  - if the `dust_balance_on_holder_tx` at the new `dust_buffer_feerate` is superior to the `max_dust_htlc_exposure_msat`:
+    - MAY fail the channel
```
```suggestion
    - SHOULD fail the channel
```

ariard replied:
See https://github.com/lightningnetwork/lightning-rfc/pull/919#discussion_r721767658

TheBlueMatt commented on 02-peer-protocol.md:
I don't think the spec needs to be this prescriptive. This is what we do in LDK, but others can/should do other things. I think we the spec should just talk generally about `feerate_per_kw` and note that in calculating you SHOULD provide some buffer for future feerate increases.

t-bast commented:
> Now, this amount has not been defined anywhere! I am reluctant to add YA random value send on connection (and it won't help existing deployments), so perhaps we define it as 0.5% of the channel capacity, and gate it on a feature bit? This may make tiny channels unusable: if we care, the language above should be changed to always allow a single dust HTLC.

It's not only tiny channels, if the feerate rises too much it becomes a real issue for a lot of channels.
At 50 sat/byte the trim threshold is around 9 000 sats, so any channel below 1 800 000 sat won't accept any HTLC.
At 75 sat/byte the treshold is around 13 000 sats, so any channel below 2 600 000 sat won't accept any HTLC.
And we've seen sustained feerates as high as that less than a year ago...

Unless we add a new rule as you suggest to allow at least one HTLC.

However, since we're all migrating to `anchor_outputs_zero_fee_htlcs` soon :tm: I'm not sure it's worth bike-shedding this too much. We won't have this issue since dust thresholds won't depend on the feerate anymore. It also removes the `update_fee` issue completely (and it is IMO the most problematic one because it leads to channels closing). At that point we only have to sometimes fail incoming htlcs instead of relaying them, which is a useful mechanism to implement anyway for throttling / mitigating channel jamming.

I think this parameter really should be configurable by each node operator based on their risk aversion and their channel size.
If I have a 10 BTC channel, I really don't want to risk burning 50 mbtc in dust (0.5%), I want to be able to set it to a much lower value.

I don't think it's worth communicating to your peers either, since it's harmless when they send you HTLCs that overflow your tolerance, they simply end up being failed. And I do see node operators updating this value over time (maybe as they build trust with their peer) which means it would need to allow updates, which is a pain.

ariard commented:
Finally updated at 0ed8805.

The level of dust exposure is recommended to be evaluated against a node-setting `max_dust_htlc_exposure_msat`, instead to introduce a new channel parameter. . The prescriptive description of `dust_buffer_feerate` as been removed, as commented [before](https://github.com/lightning/bolts/pull/919#discussion_r721867203) and is deferred to implementors.

I think few open review questions :
- describe more precisely the dust balance on holder/counterparty transactions ?
- make it more clear the mitigation distinction between pre/post-`option_anchors_zero_htlc` channels ?

t-bast commented on 02-peer-protocol.md:
```diff
@@ -797,6 +797,56 @@ A fulfilling node:
 
+### Bounding exposure to trimmed in-flight HTLCs: `max_dust_htlc_exposure_msat`
+
+When a HTLC is present in a channel is below the "trimmed" threshold in [BOLT3 #3](03-transactions.md),
```
nit:
```suggestion
When an HTLC in a channel is below the "trimmed" threshold in [BOLT3 #3](03-transactions.md),
```

t-bast commented on 02-peer-protocol.md:
nit:
```suggestion
To mitigate this scenario, a `max_dust_htlc_exposure_msat` can be applied at
```

t-bast commented on 02-peer-protocol.md:
nit: if "the" HTLC's (...)

And same for the other requirements below. Technically we should probably say "a hash-time-lock contract" but the spec uses "an HTLC' everywhere, so we should be consistent.

t-bast commented on 02-peer-protocol.md:
```diff
@@ -1127,6 +1186,25 @@ A receiving node:
+      - MAY fail the channel
+
+
+For pre-`option_anchors_zero_fee_htlc` channels, the second-stage HTLC transaction
+fee is function of the negotiated `feerate_per_kw`. To avoid a node being
+significantly more exposed to trimmed balance in case of feerate increases when
+several HTLCs are pending which are near the dust limit, it is recommended to
+calculate that `feerate_per_kw` with some buffer.
```
nit: I'd remove this duplicated paragraph (already there above).

t-bast reviewed (commented):
A few nits. I like how this version is structured, but some of the terminology is quite different from the rest of the spec, I'll give it a shot to propose a patch to harmonize the wording with the rest of the spec.

ariard commented:
Thanks for the review. Fixed all comments, if you have suggestions for the wording happy to take them!

t-bast commented:
I created a commit [here](https://github.com/t-bast/bolts/commit/7e3eb756c51fa6798d4fa934726d1781edf1ed19) with many cosmetic changes (use same terminology as the existing spec, rephrase some sentences that were unclear to me, etc). Most of those are very subjective, so feel free to judge whether it's better or not!

ariard commented:
Many thanks for the edit. Updated the local branch with it at 5fb0b5c.
Still happy if we can get in, as we might have to reconsider our per-implementation `max_dust_htlc_exposure_msat` if something like "ephemeral anchors" gets in during the coming years: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-October/021046.html It might change the attacker set of the dust exposure.

t-bast reviewed (commented):
ACK https://github.com/lightning/bolts/pull/919/commits/5fb0b5ce66e5c6a20ea7a01db942dcc7f24ee1c8

ariard commented:
Updated at 219adca

Reviewed back the recommendations against LDK logic as part of our `option_anchors_zero_fee_htlc_tx` output support. Added new precision in consequence, especially when receiving `update_fee`, there is no need to reject the message if `option_anchors_zero_fee_htlc_tx` has been negotiated.

t-bast commented on 02-peer-protocol.md:
```diff
@@ -841,7 +841,9 @@ dust exposure. The exact value used is a matter of node policy.
 the `feerate_per_kw` may trim multiple htlcs from commitment transactions,
-which could create a large increase in dust exposure.
+which could create a large increase in dust exposure. If `option_anchors_zero_fee_htlc_tx`
+has been negotiated, the HTLC feerate is implied to be 0, the checks should
+still be applied as a node can be exposed due to the sum of the trimmed HTLCs.
```
This sentence is quite unclear to me: which checks are you talking about? I don't think this new sentence is necessary at all to be honest, the requirements above already apply to all commitment types, and this paragraph only says that without `option_anchors_zero_fee_htlc_tx` you should be extra cautious when the feerate changes.

ariard replied:
Removed this sentence on latest commit `7705356`, if we think it doesn't add clarification.

ariard commented:
Updated at 7705356.

If we think there is no more value in adding a clarification in the specification about dust exposure, I can close it. On the other-side, while looking on the validating lightning signer code recently, it turns out to be affected by the same security issue (https://lists.linuxfoundation.org/pipermail/lightning-dev/2023-January/003829.html) so I wouldn't be surprised people writing that type of auxiliary lightning softwares to screw up again, if not warned.

t-bast reviewed (approved):
ACK https://github.com/lightning/bolts/pull/919/commits/77053564e66f83d47813d82de39fa3eef1824b1a

I do think there is value in merging this PR since:

- this is a non-trivial issue that implementers should know about: we help new protocol developers ramp up more quickly by mentioning this explicitly in the spec
- we've been able to reduce this to a small enough change that it doesn't add too much clutter to the existing spec

But I believe we should have an ACK from `cln` or `lnd` to move forward with adding that change.

ariard commented:
I’ll close this PR. Like said I know where are the flaws, not my issue if the wider ecosystem forgets about them.

Roasbeef commented:
This was brought up in the latest spec meeting (re better go-to-chain heuristics factoring in chain unit economics): https://github.com/lightning/bolts/issues/1082

ariard commented:
> This was brought up in the latest spec meeting (re better go-to-chain heuristics factoring in chain unit economics): https://github.com/lightning/bolts/issues/1082

Good I think this is just missing one more review to be merged on top of the one of t-bast. On the long-term, I think transaction-relay policy rules could be relaxed to have dust outputs being spend in a single package a la ephemeral anchor.

niftynei commented:
Spec meeting notes (#1098): pending review from @rustyrussell

rustyrussell commented:
There are some nits I could pick (the naming makes it sound like a field on the wire, rather than an internal threshold), but it's basically sound.  I like that it has fewer requirements with zero-fee-anchors, too.

Ack 77053564e66f83d47813d82de39fa3eef1824b1a

ariard commented:
If you have specific nits feel free to let them, I’ll address them or please add them with a follow-up PR.

Happy to add the recommendations we’re working for fee-bumping reserves management post-anchor we’re working for LDK, once we’re good with dust exposure here.

ProofOfKeags commented:
Forgive me if I'm missing something, but this doesn't seem like a protocol change so much as a policy recommendation? The only thing I can imagine leaking into the protocol would be an error message that basically said it was overallocated on htlc exposure but I would imagine that would be (and should be) classified as a temp channel failure.

If it is a policy rec, I would like to second what @rustyrussell said and at least "unquote" the term, if not move it to a different section. That said, I do see how we already have precedence for this with the "`cltv_expiry_delta` selection" section.

ariard commented:
> That said, I do see how we already have precedence for this with the "cltv_expiry_delta selection" section.

Note here the distinction between mechanism vs policy is maintained, as there is no mandatory selection of the value of `max_dust_htlc_exposure`, only laying out the “authorization” mechanism by which the per-channel dust exposure can be controlled.

Such authorization mechanism deployment alters the protocol behavior, and therefore qualifies as protocol change from my viewpoint. This is unclear to me what you qualify as a “pure” protocol change, as you note historically in the BOLT we have documented some variables selection, e.g you have other example such as fragmentation of punishment transaction post-anchor: https://github.com/lightning/bolts/blob/master/05-onchain.md#rationale-6
